### ISSUE TITLE
Optuna Suggests the Same Parameter Values in a lot of Trials (Duplicate Trials that Waste Time and Budget)

### ISSUE THREAD

**Issue by deathes** [optuna/optuna#2021]
Optuna TPESampler and RandomSampler try the same suggested integer values (possible floats and loguniforms as well) for any parameter more than once for some reason. I couldn't find a way to stop it from suggesting same values over over again. Out of 100 trials quite a few of them are just duplicates. Unique suggested value count ends up around 80-90 out of 100 trials. If I include more parameters for tuning, say 3, I even see all 3 of them getting the same values a few times in 100 trials.

It's like this:

[I 2020-11-14 14:44:05,320] Trial 8 finished with value: 45910.54012028659 and parameters: {'min_data_in_leaf': 75}. Best is trial 4 with value: 45805.19030897498.
[I 2020-11-14 14:44:07,876] Trial 9 finished with value: 45910.54012028659 and parameters: {'min_data_in_leaf': 75}. Best is trial 4 with value: 45805.19030897498.
[I 2020-11-14 14:44:10,447] Trial 10 finished with value: 45831.75933279074 and parameters: {'min_data_in_leaf': 43}. Best is trial 4 with value: 45805.19030897498.
[I 2020-11-14 14:44:13,502] Trial 11 finished with value: 46125.39810101329 and parameters: {'min_data_in_leaf': 4}. Best is trial 4 with value: 45805.19030897498.
[I 2020-11-14 14:44:16,547] Trial 12 finished with value: 45910.54012028659 and parameters: {'min_data_in_leaf': 75}. Best is trial 4 with value: 45805.19030897498.

Example code below: 

    def lgb_optuna(trial):

        rmse = []

        params = {
            "seed": 42,
            "objective": "regression",
            "metric": "rmse",
            "verbosity": -1,
            "boosting": "gbdt",
            "num_iterations":  1000,
            'min_data_in_leaf':  trial.suggest_int('min_data_in_leaf', 1, 100)
        }

        cv = StratifiedKFold(n_splits=5, random_state=42, shuffle=False)
        for train_index, test_index in cv.split(tfd_train, tfd_train[:,-1]):
            X_train, X_test = tfd_train[train_index], tfd_train[test_index]
            y_train = X_train[:,-2].copy()
            y_test = X_test[:,-2].copy()
            
            dtrain = lgb.Dataset(X_train[:,:-2], label=y_train)
            dtest = lgb.Dataset(X_test[:,:-2], label=y_test)
        
            booster_gbm = lgb.train(params, dtrain, valid_sets=dtest, verbose_eval=False)

            y_predictions = booster_gbm.predict(X_test[:,:-2])
            final_mse = mean_squared_error(y_test, y_predictions)
            final_rmse = np.sqrt(final_mse)
            rmse.append(final_rmse)
    
         return np.mean(rmse)

    study = optuna.create_study(sampler=TPESampler(seed=42), direction='minimize') 
    study.optimize(lgb_optuna, n_trials=100)

**Comment by crcrpar** [comment]
Desirably, if the suggested parameters are the ones already suggested, the trial should be skipped for faster hyperparameter optimization, right?

**Comment by deathes** [comment]
> Desirably, if the suggested parameters are the ones already suggested, the trial should be skipped for faster hyperparameter optimization, right?
> 
> Reproduced output: https://gist.github.com/crcrpar/aba308a1350bb4986276a6c87cf256cb

Exactly. If the trial has already been suggested, it should be skipped for faster hyperparameter optimization.

**Comment by crcrpar** [comment]
By its nature, TPESampler tends to sample similar values with the number of trials increasing since it narrows the search space based on Bayesian optimization with the number of trials increasing. So I think it's a bit hard to avoid the same values suggested.

That being said, we can deal with this case by checking the suggested values in the `objective` function and if they aren't new ones, terminate the trial.

```python
def objective(trial):
    some_param = trial.suggest_int("some_param", 0, 100)
    if some_param in param_history["some_param"]:
        raise optuna.exceptions.TrialPruned()
    ...
    # define your evaluation 
    return value
```
The full code with this hack and its result is available: https://gist.github.com/crcrpar/c6d6fdf8112280654884353d6e68f6bb.

Note that this solution might not be general enough to apply to other cases.

@HideakiImamura would have some better ideas.

**Comment by hvy** [comment]
Yes, and pruned past trials are treated differently for different samplers so it might or might not work. You could also just fail the trials by raising other exceptions or return an invalid value such as `None`. The issue is that the sampler is likely to sample similar values in the next trial. `TPESampler` for instance would. As for the `TPESampler` and the `RandomSampler`, these behaviors are by design.

Just based on the code snippet above, let me just clarify that in case you're working with a discrete search space and your trial budget (`n_trials`) is larger than or equal to the size of the space, you could also use the [`GridSampler`](https://optuna.readthedocs.io/en/stable/reference/generated/optuna.samplers.GridSampler.html). You'll avoid reevaluating objective functions with the same configurations (although it's not 100% guaranteed in distributed settings due to timing issues).

However, to properly address this issue we probably have to work on a new sampler based on a different algorithm.

**Comment by bfs15** [comment]
If anyone cares enough about this in the meantime, I made a class for this.
Prunes trials with all parameters equal to some existing past trial. You also can allow repeats if you want, and instead of pruning, you can get the average of these past trials, useful for nondeterministic objective functions.
Duplicates of running trials can happen when multiprocessing, so I tried to take care of that too. It's possible a running trial can be pruned in the future, and I don't want trials to be registered as repeats of pruned ones because of the "average value of past trials" thing. So I register them as unfinished_repeats and check up on them on future prunes.

If you really want to prune repeats of pruned trials (or `FAIL` ones) you can add the PRUNED state in the `should_compare_states` constructor argument.
If you want to allow some repeats use the `repeats_max` argument, the default is no repeats (0).

https://gist.github.com/bfs15/24045ab5e8ad007b4a09f708adfe359f

**Comment by not522** [comment]
Re-open is welcome.
By the way, it is difficult to exclude duplicated trials as the default behavior. In the following case, parameters should be checked after all suggestions. We need to check them manually as above.

```python
import optuna

def objective1(trial):
    x = trial.suggest_int("x", 0, 0)
    return x

def objective2(trial):
    x = trial.suggest_int("x", 0, 0)
    # Here, it looks the same parameters but should not be pruned.

    y = trial.suggest_int("y", 0, 0)
    # We should check after all parameters have been suggested.

    return x + y

study = optuna.create_study()
study.optimize(objective1, n_trials=1)
study.optimize(objective2, n_trials=1)
```

**Comment by cowwoc** [comment]
@not522 Maybe we could have a built-in pruner that rejects duplicates and rolls back the number of trials by one automatically.

**Comment by jmrichardson** [comment]
+1 vote to add this functionality

**Comment by cowwoc** [comment]
Thinking about this further, is this really a problem? While it is quite possible that the same parameter value will get suggested multiple times it seems extremely unlikely that *all* parameter values will be equal. So long as at least one parameter is different, isn't it worth running the trial?

**Comment by jmrichardson** [comment]
> 
> 
> Thinking about this further, is this really a problem? While it is quite possible that the same parameter value will get suggested multiple times it seems extremely unlikely that _all_ parameter values will be equal. So long as at least one parameter is different, isn't it worth running the trial?

IMO, yes this is a problem worth resolving. Here are my concerns:  

- I have many use cases where Optuna optimizes a single parameter.  Therefore, the trials are often repeats of the same value.
- Having to code a workaround to handle repeats shifts the burden to coding logic which introduces complexity
- Unfortunately, the way my project is coded, the parameter suggestions are contained in an eval method which makes it impossible to intercept prior to the function executing.  Therefore, the function will execute and consume resources.

I don't see any value in Optuna running the same trial multiple times. In my humble opinion, it just makes sense to handle within the lib.

**Comment by gpogorzelski** [comment]
I think @jmrichardson is right. In my work, after training for many iterations I collect say 10 models with best performance and evaluate them further (do they cluster data? etc.). out of the 10 with best performance I select the final candidate. With a lot of trials I tend to get all 10 models with the same parameters. Obviously I implemented a workaround for this, but I believe it should be handled within the library.

**Comment by blazespinnaker** [comment]
This one is a bit tricky to do with distributed optuna as we'll need to query from storage what the historical parameters are.  I was using it for feature selection, and it's a very serious problem as it gets rather stuck in exploitation and wastes a crazy amount of cycles.  I was very surprised to see this behavior and had assumed that the sampler naturally handled this use case.

I think a good way to handle this might be by getting a stable hash of the parameters and putting it into the storage.

**Comment by blazespinnaker** [comment]
![Screenshot 2022-08-16 5 20 46 PM](https://user-images.githubusercontent.com/2188809/185007286-d911a3ed-14d3-4fa2-99dd-ce00166cedf4.png)

Around 30% of the trials are being duplicated.

**Comment by not522** [comment]
In the black-box optimization, we can suppose "noisy" functions, i.e., including randomness in the objective values (E.g., https://arxiv.org/abs/2105.08195). For these problems, re-evaluating the objective function may be informative. Please note samplers don't know whether the objective function is noisy or not.
When the objective function is deterministic, samplers expect it returns the same value as previous trials. It can implement as in the following example.

```python
import optuna
from optuna.trial import TrialState

def objective(trial):
    x = trial.suggest_int("x", 0, 1)
    y = trial.suggest_int("y", 0, 1)

    for previous_trial in trial.study.trials:
        if previous_trial.state == TrialState.COMPLETE and trial.params == previous_trial.params:
            print(f"Duplicated trial: {trial.params}, return {previous_trial.value}")
            return previous_trial.value

    return x + y

study = optuna.create_study()
study.optimize(objective, n_trials=10)
```

**Comment by teneon** [comment]
This "bug" is the first thing I have noticed when i first started using/testing Optuna. If you are optimizing just one or two parameters, it will surely select exactly the same two parameter multiple times in your trials. Imagine, you are training a huge neural network and one trial takes a lot of time. Now out of 100 trials, you have repeated e.g. 30% of them. That is such a waster of time and resources. Hopefully it gets implemented/fixed soon.

**Comment by louis-she** [comment]
> In the black-box optimization, we can suppose "noisy" functions, i.e., including randomness in the objective values (E.g., https://arxiv.org/abs/2105.08195). For these problems, re-evaluating the objective function may be informative. Please note samplers don't know whether the objective function is noisy or not. When the objective function is deterministic, samplers expect it returns the same value as previous trials. It can implement as in the following example.
> 
> ```python
> import optuna
> from optuna.trial import TrialState
> 
> def objective(trial):
>     x = trial.suggest_int("x", 0, 1)
>     y = trial.suggest_int("y", 0, 1)
> 
>     for previous_trial in trial.study.trials:
>         if previous_trial.state == TrialState.COMPLETE and trial.params == previous_trial.params:
>             print(f"Duplicated trial: {trial.params}, return {previous_trial.value}")
>             return previous_trial.value
> 
>     return x + y
> 
> study = optuna.create_study()
> study.optimize(objective, n_trials=10)
> ```

Parameters Tunning makes much more sense when the `objective function` is deterministic, cause if it is not, why not just run the same code repeatedly without any auto tunning tools?

**Comment by michaelwayman** [comment]
+1.. I am here because I was running a study to determine which initializer worked best. Each trial takes quite a while. When I saw that all 3 trials ran the same initializer I was very confused. My assumption was that optuna would use a separate categorical initializer each trial, especially considering it was the only parameter.
The comments here even gave a few good suggestions. I will open a PR myself if I think it has a chance of being merged.

@Optuna is there a solution I could open a PR for that you would find acceptable?

**Comment by not522** [comment]
If you want to try all parameters, I recommend `BruteForceSampler`.
https://optuna.readthedocs.io/en/v3.1.0-b0/reference/samplers/generated/optuna.samplers.BruteForceSampler.html

**Comment by blazespinnaker** [comment]
@not522 I think answers like that in this thread are more likely to generate heat rather than light.  Maybe a more basic explanation giving folks the benefit of the doubt of why you think this isn't a relevant issue might be helpful.  Clearly there is a lot of interest here, but perhaps there is a clear and well laid out argument to be made as why there shouldn't be.

**Comment by blazespinnaker** [comment]
I agree with your conclusion most def, though I am but unclear on the wording of the 1 and 2 in relation to that.  3 makes sense.

**Comment by blazespinnaker** [comment]
I agree, most definitely.

**Comment by not522** [comment]
@blazespinnaker Sorry if I have offended you. I just wanted to provide a workaround because this issue is hard to solve.

I try to explain why this "bug" is so difficult, even though it seems very simple. Of course, I agree the current behavior is non-intuitive. If anyone have a great idea to solve my concerns, I would be very interested in hearing about it.

I have three concerns about implementing this feature. First, when do samplers determine the search space? Second, how to avoid duplicate sets? Third, what happens when all sets are evaluated?

### When do samplers determine the search space?

As I wrote previously, it is difficult to check for duplicate parameters. Optuna supports dynamic search space, so the sampler cannot determine the search space until the trial completes.

```python
import optuna

def objective1(trial):
    x = trial.suggest_int("x", 0, 0)
    return x

def objective2(trial):
    x = trial.suggest_int("x", 0, 0)
    # Here, it looks the same parameters, but they are not.

    y = trial.suggest_int("y", 0, 0)
    # We should check after all parameters have been suggested.

    return x + y

study = optuna.create_study()
study.optimize(objective1, n_trials=1)
study.optimize(objective2, n_trials=1)
```

It is reasonable to estimate the search space is the same as in previous trials. Optuna provides this feature named relative sampling. But it has a limitation. It cannot handle dynamic search space, for example, branching search space and changing the objective function. So, only if all parameters are suggested in all trials, and the suggestion ranges are fixed, we have room to resolve this issue using relative sampling. (Note: The group decomposed search space may relax the limitation.)

### How to avoid duplicate sets?

If the sampler knows the search space in some way, we have a chance to remove duplicate parameters, but how? One idea is to resample parameters if they are duplicated from previous trials (like rejection sampling). However, this method can take a long time to suggest new parameters when the search converges.

### What happens when all sets are evaluated?

When the objective function is fixed -- Optuna does not assume it, but I ignore it here --, samplers can notice all sets are evaluated. `BruteForceSampler` realizes it and automatically stops the study. But as the default behavior for the default sampler, what should happen? Let us consider the following example.

```python
import optuna

def objective(trial):
    return trial.suggest_int("x", 1, 3)

study = optuna.create_study()
study.optimize(objective, n_trials=10)
```

In this case, all sets of parameters can be evaluated by three trials, but the user specifies `n_trials=10`. I think ignoring `n_trials=10` is also non-intuitive. Users can stop the optimization using `BruteForceSampler` / `GridSampler` or a callback function as an argument of `optimize`. Either way, the user's intent needs to be explicitly stated.

### My conclusion

This issue is complex to solve, despite appearances. I think this feature is difficult to introduce as the default behavior and recommend workarounds using `BruteForceSampler` or the previous trial's value.

```python
import optuna
from optuna.trial import TrialState

def objective(trial):
    x = trial.suggest_int("x", 0, 1)
    y = trial.suggest_int("y", 0, 1)

    for previous_trial in trial.study.trials:
        if previous_trial.state == TrialState.COMPLETE and trial.params == previous_trial.params:
            print(f"Duplicated trial: {trial.params}, return {previous_trial.value}")
            return previous_trial.value

    return x + y

study = optuna.create_study()
study.optimize(objective, n_trials=10)
```

As @blazespinnaker suggested, encouraging exploration can relax the problem. I am interested in this but need a discussion of the TPE algorithm.

(repeat) Of course, I agree the current behavior is non-intuitive. If anyone have a great idea to solve my concerns, I would be very interested in hearing about it.

**Comment by contramundum53** [comment]
Does PR #4352 solve this problem?

**Comment by contramundum53** [comment]
I think it is important for TPE algorithm to "see" the duplicate trials to converge to the desired solution. However, if you know that the objective function is deterministic, you could memoize it. #4352 provides a helper function for it.

**Comment by cowwoc** [comment]
@contramundum53 if I understand correctly, you are caching the output of the last input. I would expect caching of multiple values. Ideally, I'd want to cache the X most frequent inputs where X could be infinity.

**Comment by contramundum53** [comment]
@cowwoc No, optuna keeps all the history of previous evaluations, and #4352 simply searches the entire history for exact match. This is equivalent to X=infinity case.

**Comment by contramundum53** [comment]
Sorry if the name `last_value_if_duplicate` is misleading, the name indicates that if there happens to be multiple same inputs (this is still possible with distributed optimization), it picks the last one.

**Comment by cowwoc** [comment]
> Sorry if the name `last_value_if_duplicate` is misleading, the name indicates that if there happens to be multiple same inputs (this is still possible with distributed optimization), it picks the last one.

Right, that's what confused me. Maybe it should be called `last_duplicate_value`? And actually, if this mechanism is in play then why are duplicate values getting recorded at all? Shouldn't only the first duplicate get logged.

**Comment by contramundum53** [comment]
In distributed environment, it is possible that multiple trials with the same parameters are sampled "at the same time" before the results are `tell`-ed. With some more engineering, we could also detect duplicate `RUNNING` trials and fail the second ones, but I haven't implemented that in the above PR.

**Comment by camall3n** [comment]
This is a huge issue for distributed optimization. I'm seeing 10 workers, each running identical params, and each getting the same result, then all 10 workers sample another set of identical params and run those with the same result, etc.

**Comment by camall3n** [comment]
The situation I described was with `seed=None`, but I take your point. I was able to resolve it by making multiple `create_study()` calls, one in each worker, and initializing the sampler in each worker with a different seed.

Here's the gist: https://gist.github.com/camall3n/4b0272e547b0e5df05777ebed446b69f

**Comment by gabriel-vanzandycke** [comment]
I drives me crazy that same point in the hyper-parameter space is sampled multiple times (see runs 18, 20, 21 and 22) 
```
|     17 | [0.31024930747922436, 0.9739130434782609] | 2023-01-27 15:52:51 | 2023-01-27 15:59:02 | 0:06:11.199076 | {'min_distance_cm': 100, 'min_distance_px': 50}  | COMPLETE |
|     18 | [0.55625, 0.8811881188118812]             | 2023-01-27 15:53:00 | 2023-01-27 16:01:19 | 0:08:18.946978 | {'min_distance_cm': 50, 'min_distance_px': 100}  | COMPLETE |
|     19 | [0.31024930747922436, 0.9739130434782609] | 2023-01-27 15:55:47 | 2023-01-27 16:01:53 | 0:06:06.252989 | {'min_distance_cm': 100, 'min_distance_px': 50}  | COMPLETE |
|     20 | [0.55625, 0.8811881188118812]             | 2023-01-27 15:59:04 | 2023-01-27 16:04:59 | 0:05:55.289485 | {'min_distance_cm': 50, 'min_distance_px': 100}  | COMPLETE |
|     21 | [0.55625, 0.8811881188118812]             | 2023-01-27 15:59:39 | 2023-01-27 16:05:55 | 0:06:16.187209 | {'min_distance_cm': 50, 'min_distance_px': 100}  | COMPLETE |
|     22 | [0.55625, 0.8811881188118812]             | 2023-01-27 16:01:21 | 2023-01-27 16:09:17 | 0:07:55.919076 | {'min_distance_cm': 50, 'min_distance_px': 100}  | COMPLETE |
|     23 | [0.20555555555555555, 0.9568965517241379] | 2023-01-27 16:01:54 | 2023-01-27 16:07:51 | 0:05:56.547381 | {'min_distance_cm': 50, 'min_distance_px': 25}   | COMPLETE |
```
I'm using `GridSampler({"min_distance_cm": range(50, 101, 25), "min_distance_px": range(25, 101, 25)})` as [suggested by hvy](https://github.com/optuna/optuna/issues/2021#issuecomment-729368756).

**Comment by contramundum53** [comment]
@gabriel-vanzandycke Which version of optuna are you using? This problem is supposed to be fixed in #2783 (released in v2.9), except for a few trials in the end.

**Comment by classner** [comment]
Hi!

Great discussion! I would very much argue we need a solution for this problem, and ideally within Optuna. I totally see that there is no 'one-size-fits-all' solution, and users will have to make a choice about this (with sensible defaults ideally). Ax for example, has a flag that allows to specify whether the experiment should be regarded as deterministic.

It is important to find a solution because the phenomenon of repeated experiment values is quite prevalent from what I can see, and it burns very real time and money. I am using Optuna through the plugin for hydra, hence do not have an overview of trials to fall back to recent results (the experiments themselves are not aware of the hyperparameter optimization process), so a configurable solution from within Optuna would be amazing to have!

**Comment by AliAkhgari** [comment]
Use this pruner:

```import optuna
from optuna import logging
from optuna.pruners import BasePruner

_logger = logging.get_logger(__name__)


class DuplicateIterationPruner(BasePruner):
    """
    DuplicatePruner

    Pruner to detect duplicate trials based on the parameters.

    This pruner is used to identify and prune trials that have the same set of parameters
    as a previously completed trial.
    """

    def prune(
        self, study: "optuna.study.Study", trial: "optuna.trial.FrozenTrial"
    ) -> bool:
        completed_trials = study.get_trials(states=[optuna.trial.TrialState.COMPLETE])

        for completed_trial in completed_trials:
            if completed_trial.params == trial.params:
                return True

        return False

**Comment by tniveej** [comment]
> Hi!
> 
> Great discussion! I would very much argue we need a solution for this problem, and ideally within Optuna. I totally see that there is no 'one-size-fits-all' solution, and users will have to make a choice about this (with sensible defaults ideally). Ax for example, has a flag that allows to specify whether the experiment should be regarded as deterministic.
> 
> It is important to find a solution because the phenomenon of repeated experiment values is quite prevalent from what I can see, and it burns very real time and money. I am using Optuna through the plugin for hydra, hence do not have an overview of trials to fall back to recent results (the experiments themselves are not aware of the hyperparameter optimization process), so a configurable solution from within Optuna would be amazing to have!

I agree with this, I use Optuna through Ray and find it hard to implement such changes as well.

**Comment by contramundum53** [comment]
@tniveej I think you can write a sampler wrapper that removes duplicate parameters, which you can then plug in to the `sampler` argument of the Ray's OptunaSearch class.
I'll write one when I have time.
(For hydra, you might need to add path to PYTHONPATH so that hydra sees your sampler.)

**Comment by drozzy** [comment]
In RL you have stochastic policies and stochastic enviornments.
The value of objective function can vary wildly between runs.

**Comment by Loccret** [comment]
Hello everyone,

I understand that mathematically, Bayesian optimization can converge to a local optimum. Does this also occur in Optuna?
Additionally, if convergence is observed, say the same hyperparameters are chosen multiple times, around 10 instances, how might I switch to random sampling to potentially locate a more optimal point than the current best?

**Comment by contramundum53** [comment]
TPE can converge to a local optimum, but it can and actually very often sample the same hyperparameters multiple time long before convergence (if all parameters are discrete).
You can switch to random sampling after convergence, but I think 10 trials is too early to switch.

**Comment by contramundum53** [comment]
A descent implementation of GP would do that much less likely. However, currently `BoTorchSampler` would not perform well if you have a categorical parameter.

**Comment by contramundum53** [comment]
`GPSampler(deterministic_objective=True)` should solve this problem, although it may still suggest duplicate trials due to precision loss.

**Comment by oasidorshin** [comment]
Recently opened issue again (didn't know about this one), and wanted to ask why the solution is still user-added boilerplate code from the FAQ? It would be much easier to implement this as another parameter for the sample, i.e. previously mentioned `last_value_if_duplicate` (https://github.com/optuna/optuna/pull/4352).

I want to mention that in machine learning there is a huge problem of copy-pasted boilerplate code for everything, and right now instead of simple parameter users have to remember to insert several rows into their objective function.

I suggest to reconsider this decision, with many users asking for built-in solution in this issue. Thank you in advance! @contramundum53 @nzw0301

**Comment by not522** [comment]
Thank you for your feedback. We see this issue as a high-demand problem with a lot of interest. However, how to resolve this issue has taken a lot of discussion and we have yet to do anything other than put a work-around in the FAQ. #4352 is an option, but instead of being easier to implement, it compromises the simplicity of the API.
Please feel free to send us your requests, but please understand that we are very careful about adding APIs so as not to lose the simplicity of Optuna.

**Comment by deadsoul44** [comment]
Shameless plug:
https://github.com/perpetual-ml/perpetual

**Comment by JuniorDurand** [comment]
I was experiencing the same problem and solved it with the function below, which returns the same value as a previous trial.

```python
def check_repeated_trial(trial):
  optuna_study = trial.study
  

  for past_trial in optuna_study.get_trials():
    if past_trial.number == trial.number:
      continue

    past_params = past_trial.params
    repeated_trial = True
    
    for key in trial.params:
      if key in past_params and trial.params[key] != past_params[key]:
        repeated_trial = False
        break 
    
    if repeated_trial is True:
      print(f"past Trial: {past_trial.number}, current Trial: {trial.number}")
      print(f"Params of past Trial: {past_params}")
      print(f"Params of current Trial: {trial.params}")
      return past_trial.value

  return None
  ```

**Comment by michabaron98** [comment]
+1
